guide/adopting-the-configuration-cache.md

```diff
@@ -0,0 +1,245 @@
+
+
+## Solving Configuration Cache problems
+
+Let's look at an example of 
+
+### Before:
+
+- `ZstdCompressTask` is already a gradle-managed type
+- `ZstdCompressor` is a POJO.
+- `ZstdCompressor`  starts an external process, causing the configuration cache to fail: **external process started**
+- Furthermore, `ZstdCompressTask` is calling `getProject()` at build time, causing another configuration cache failure: **cannot serialze object of type `org.gradle.api.internal.project.DefaultProject`**  
+
+```java
+abstract class ZstdCompressTask extends org.gradle.api.DefaultTask {
+    @Input
+    protected abstract Property<Path> getInputFile();
+
+    @Input
+    protected abstract Property<Path> getOutputFile();
+
+    @TaskAction
+    public void compress() {
+        ZstdCompressor compressor = new ZstdCompressor(5);
+        compressor.compress(getProject(), getInputFile().get(), getOutputFile().get());
+    }
+}
+
+class ZstdCompressor {
+    private int compressionLevel;
+
+    public ZstdCompressor(int compressionLevel) {
+        this.compressionLevel = compressionLevel;
+    }
+
+    public void compress(Project project, Path inputFileRelative, Path outputFileRelative) {
+        Path projectDir = project.getProjectDir().toPath();
+        Path inputPath = projectDir.resolve(inputFileRelative);
+        Path outputPath = projectDir.resolve(outputFileRelative);
+
+        try {
+            ProcessBuilder pb = new ProcessBuilder(
+                    "zstd", "-" + compressionLevel, inputPath.toString(), "-o", outputPath.toString());
+            Process process = pb.start();
+            int exitCode = process.waitFor();
+            if (exitCode != 0) {
+                throw new RuntimeException("zstd compression failed with exit code: " + exitCode);
+            }
+        } catch (IOException | InterruptedException e) {
+            throw new RuntimeException("Failed to execute zstd compression", e);
+        }
+    }
+}
+```
+
+### Step 1: Fixing "external process started"
+
+- To safely do the external call, we need to use `ExecOperations` in `ZstdCompressor`
+- To use `ExecOperations`, we can first make `ZstdCompressor` a Gradle-managed type, then inject `ExecOperations`
+- To make `ZstdCompressor` a Gradle-managed type, we make it abstract. It now needs to be instantiated with `ObjectFactory::newInstance`
+- To get an `ObjectFactory`, we inject it into `ZstdCompressTask`, which is conveniently already a Gradle-managed type
+
+```java
+abstract class ZstdCompressTask extends org.gradle.api.DefaultTask {
+    @Input
+    protected abstract Property<Path> getInputFile();
+
+    @Input
+    protected abstract Property<Path> getOutputFile();
+
+    @Inject
+    protected abstract ObjectFactory getObjectFactory();
+
+    @TaskAction
+    public void compress() {
+        ZstdCompressor compressor = getObjectFactory().newInstance(ZstdCompressor.class, 0);
```

Comment:
The ideal would be to make this an `@Internal @Nested` but I wonder whether that's too many ideas to introduce here...

Reply:
We could, but I wanted to give the example of a regular stateful java class with non-static members and constructors with arguments. I think if people know how to deal with this case, then other simpler cases should be self-evident.

But yep @Nested is a neat trick people might appreciate. I've added a small snippet explaining the use of @Nested in the guide, let me know what you think of it.